### ISSUE TITLE
Error when running realsense_mesh.py example

### ISSUE THREAD

**Issue by nemtiax** [JeremyBYU/polylidar#4]
**Describe the bug**
When running the realsense_mesh.py script, I get an error suggesting an incorrect argument type on a call to rotate():

```Traceback (most recent call last):
  File "examples/python/realsense_mesh.py", line 196, in <module>
    main()
  File "examples/python/realsense_mesh.py", line 188, in main
    pcd = pcd.rotate(R_Standard_d400[:3, :3], center=False)
TypeError: rotate(): incompatible function arguments. The following argument types are supported:
    1. (self: open3d.open3d_pybind.geometry.Geometry3D, R: numpy.ndarray[float64[3, 3]], center: numpy.ndarray[float64[3, 1]]) -> open3d.open3d_pybind.geometry.Geometry3D

Invoked with: geometry::PointCloud with 24259 points., array([[ 1.,  0.,  0.],
       [ 0.,  0.,  1.],
       [ 0., -1.,  0.]]); kwargs: center=FalseTraceback (most recent call last):
  File "examples/python/realsense_mesh.py", line 196, in <module>
    main()
  File "examples/python/realsense_mesh.py", line 188, in main
    pcd = pcd.rotate(R_Standard_d400[:3, :3], center=False)
TypeError: rotate(): incompatible function arguments. The following argument types are supported:
    1. (self: open3d.open3d_pybind.geometry.Geometry3D, R: numpy.ndarray[float64[3, 3]], center: numpy.ndarray[float64[3, 1]]) -> open3d.open3d_pybind.geometry.Geometry3D

Invoked with: geometry::PointCloud with 24259 points., array([[ 1.,  0.,  0.],
       [ 0.,  0.,  1.],
       [ 0., -1.,  0.]]); kwargs: center=False
```

**To Reproduce**
Simply run realsense_mesh.py.

**Expected behavior**
I think the problem is that somewhere along the way, an open3d point cloud is being passed around rather than a numpy array, or vice-versa.  Obviously, the expected behavior is that it shouldn't crash, but I haven't been able to sort out where exactly the wrong type starts being used.

**Environment  (please complete the following information):**
 - OS: MacOS
 - Architecture: 2.6 GHz Intel Core i7
 - Version: 10.13.6 High Sierra

**Additional context**

**Comment by JeremyBYU** [comment]
I think I know what it is. Open 3D had a breaking change in the API in the `rotate` function from 0.9.0 to 0.10.0. I have not updated my code to work with 0.10.0. So you will have to modify the code for the new [API](http://www.open3d.org/docs/release/python_api/open3d.geometry.PointCloud.html#open3d.geometry.PointCloud.rotate) or downgrade to 0.9.0. Looks like the new `rotate` expects `center` to be a numpy array. It [used](http://www.open3d.org/docs/0.9.0/python_api/open3d.geometry.PointCloud.html#open3d.geometry.PointCloud.rotate) to take a boolean value indicating the center of the geometry should be used for rotation. 

I'm not sure what should be passed now for `center` to have the same effect.  Maybe call `get_center()` on the point cloud and pass the result to rotate? Hahha extra work. 

Also I have never run this code on MacOS so this should be interesting.

**Comment by nemtiax** [comment]
Sounds good.  I'll see what I can work out.  So far everything else has worked fine on MacOS.  The only reason I was trying to run this example is that I wanted a better way to visualize discovered planes and polygons for large point clouds.  Matplotlib doesn't really cut it once you have hundreds of thousands of points, and so I figured I'd try following along with the realsense_mesh example.  Is there a simpler API call somewhere to export detected planar faces?

**Comment by JeremyBYU** [comment]
Completely agree with you on matplotlib. I've had a hard time finding a good 3D plotting library and open3D seems to be the most straightforward and best one for me (of course with the amazing algorithms as well). Glad to hear its working so far on MacOS.

As for the API. I am not entirely sure what part of the API you are referring and what form (i.e. data type) of "planar faces" you want exported. The raw output of Polylidar3D (which I call `planes` and `polygons`) is definitely not that immediately useful for some form of export. The `planes`  contain the triangle *indices* in the mesh that are flat and connected. The polygons contain the linear rings with point *indices*. I convert the polygons to shapely geometries that most people (and myself) are comfortable with. Are you interested in the triangle sets (planes) or the polygons? What format exactly are you interested in for an "export".

I have a repo dedicated to Polylidar3D integration with [Intel RealSense cameras](https://github.com/JeremyBYU/polylidar-realsense/tree/master). I would recommend looking at that one as well. It can project the polygons into the image and you can export them as video if you desire.

**Comment by nemtiax** [comment]
Well, I was figuring that if I couldn't get matplotlib to behave, maybe I could export the polygons and corresponding plane segments in a form I could view in something like MeshMixer.  But that's just for the purpose of checking whether my outputs look reasonable.  I was thinking if I had the set of vertices on the edge of the polygon for a plane, then I could write that to a .ply file or similar.  Maybe an easy thing for me to do would be to do least mean squares to fit an actual flat plane to the set of points in a detected plane, and then project the polygon onto that flat plane.

I'll have a look at the realsense repo.  Projecting the polygons onto an image also sounds like a perfectly reasonable way for me to visualize them, so maybe that's an easier way to go.

Thanks for the quick replies!

**Comment by JeremyBYU** [comment]
No problem. I am personally using Open3D to plot the flat triangle segments (color coded by [plane](https://github.com/JeremyBYU/polylidar/blob/d69d81faad10f76be40c73f86ead71093bb995fd/examples/python/util/realsense_util.py#L148-L162)) and and plot the polygons in Open3D (I plot the linear rings using a module I created called a [LineMesh](https://github.com/intel-isl/Open3D/pull/738#issuecomment-564785941). The green line represents the polygon of  my basement floor (blue triangular segment). The other colored segments were captrued that were "planar" but filtered out as polygons during [polygon post-processing](https://jeremybyu.github.io/polylidar/python_api/polylidar.polylidarutil.plane_filtering.filter_planes_and_holes.html#polylidar.polylidarutil.plane_filtering.filter_planes_and_holes) because they are too small. 

![image](https://user-images.githubusercontent.com/6628085/89532538-ee99be80-d7bf-11ea-8715-fc53e4397864.png)


I also do a similar thing of projecting the polygons onto the geometric plane. I basically rotate the point cloud to align the XY plane (I know the rotation needed from the normal of the geometric plane that represents  the planar segment). After projection I can then perform polygon post processing using shapely (shapely only operates on XY points on the plane). After post-processing I recover the 3D nature of the points and rotate back to what it as before.

Basically there are a lot of steps to get to that picture but should be able to be reproduced in `realsense_mesh.py`.

**Comment by nemtiax** [comment]
I got the plane extraction working, and everything looks good.  I'm now trying to recover other planes that are not z-aligned (i.e. are not the ground plane).  I believe I have what qualifies as an "organized point cloud" - it's a range map coming from a depth camera, but I'm not sure how I need to format that to use with extract_tri_mesh_from_organized_point_cloud.  In realsense_mesh.py, it looks like it's pulling out the depth channel of the rgbd image returned by the camera, and running extract_tri_mesh_from_float_depth.  I could revert my data to a depth map rather than a point cloud, but I'd prefer to just pass the point cloud in directly.  I haven't been able to find any examples that use extract_tri_mesh_from_organized_point_cloud, so I'm a bit stuck on what I'm supposed to pass to it.  In particular, how should I order points in my point cloud (I think in rows and then columns? Top to bottom?) and how should I represent points for which I got no return?

Thanks again for all your help!

**Comment by JeremyBYU** [comment]
These are all fantastic questions. The good news is that I have answers to all of them! The bad news is that it is clear that I have not done a great job of documenting my work. Sorry about that. Polylidar3D has so many parts (at least in my eyes) that I find it difficult to document everything. By the way here is my paper (preprint) that discusses Polyidar3D which may be of interest to you (it touches on the every question you asked about): https://arxiv.org/abs/2007.12065 . I think section 4.2 (shape of organized point cloud) and Section 6 (identifying dominant plane normals besides "z-aligned/ground plane").

First I dont have time to give a thorough or perfect answer but will attempt to update the documentation website with an example. Here are some brief answers though but expect a thorough example soon:

1. Yes you can extract surfaces that are not just the ground plane. You will need to use the Fast Gaussian Accumulator that I created to identify dominant plane normal in the scene (only works reliably with dominant planes, eg., ground and walls).
2. Yes you are describing an organized point cloud
3. You can keep it as an organized point cloud. You dont have to revert it back to depth. 
4. It is arranged into rows and columns. The origin is the top left. Missing data has to be represented as the NaN floating point value. The Shape is MXNX3. Where the 3 is xyz of the point.
5. Please see these lines in Polylidar realsense. It may be helpful - https://github.com/JeremyBYU/polylidar-realsense/blob/master/surfacedetector/capture.py#L229-L257

Sorry I had to be brief. I will work on an example (in this repo) with just and organized point cloud and extracting multiple surfaces.

**Comment by nemtiax** [comment]
Awesome, this is very helpful!  I already got the FastGaussianAccumulator working, although I was feeding it the mesh from a 2.5D triangulation, which is of course a little wonky on non-ground-aligned surfaces.  I'm sure switching to the proper mesh will be much better.  It sounds like it'll be pretty easy to get the organized point cloud working.

Your existing documentation has been extremely helpful so far!  I've been able to answer almost all of my questions with it,  just these few have slipped through.

Thanks again!

**Comment by JeremyBYU** [comment]
Here is some updated documentation for organized point clouds: https://jeremybyu.github.io/polylidar/tutorial/Python/organizeddemo.html#Extract-Dominant-Planes-From-Organized-Point

Hope this helps!

**Comment by nemtiax** [comment]
Wow, this is perfect!

Thanks!

**Comment by JeremyBYU** [comment]
PS you should update organized point filters repo as well as Polylidar3d. I made changes in both.

 I'll keep this issue open until I figure out a way to get both apis of open3d working for 'rotate'.

**Comment by JeremyBYU** [comment]
So I just implemented the [fix](https://github.com/JeremyBYU/polylidar/commit/a3254241777bea125d3aae1af124d89ffe073628) and upgraded to 0.10.0 . There seems to be a serious performance regression with this new version of open3d (0.10.0). Everything is very slow and the fps is now like 0.5 vs a smooth 60 fps before! Here is a link to the issue from another user who noticed: https://github.com/intel-isl/Open3D/issues/2157#issue-676035173

For now the API is now compatible with both methods of calling to `.rotate()`. I would recommend continuing to use open3d==0.9.0.

**Comment by JeremyBYU** [comment]
I updated the [documentation ](https://jeremybyu.github.io/polylidar/install_instructions.html)with warnings about the new version of Open3D. I am closing this issue now. Thanks for your help
